### ISSUE TITLE
Early stopping with custom metric

### ISSUE THREAD

**Issue by zzzbobml** [microsoft/LightGBM#4249]
Is there a working example on customizing early stopping? In particular, I'd be interested about getting access to the binary_logloss of all trees up to the current tree.

**Comment by StrikerRUS** [comment]
Please take a look at [`advanced_example.py`](https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py).

For the 
> Early stopping with custom metric

```py
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split

import numpy as np
import lightgbm as lgb

X, y = load_boston(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
train_data = lgb.Dataset(X_train, y_train)
valid_data = lgb.Dataset(X_test, y_test, reference=train_data)

def rmsle(preds, train_data):
    labels = train_data.get_label()
    return 'RMSLE', np.sqrt(np.mean(np.power(np.log1p(preds) - np.log1p(labels), 2))), False

bst = lgb.train(params={'objective': 'regression', 'metric': 'None'},
                train_set=train_data, num_boost_round=20,
                valid_sets=[valid_data], feval=rmsle, early_stopping_rounds=2)
```


> In particular, I'd be interested about getting access to the binary_logloss of all trees up to the current tree.

Could you please elaborate?
I think you can achieve this by writing your own callback and access info in `evaluation_result_list` field of callback environment.
https://github.com/microsoft/LightGBM/blob/aedfdd0d1e4ac570b539d17d74de1610acd099df/python-package/lightgbm/callback.py#L28-L36
Or slightly re-write `record_evaluation` callback
https://github.com/microsoft/LightGBM/blob/aedfdd0d1e4ac570b539d17d74de1610acd099df/python-package/lightgbm/callback.py#L75-L76
Refer to as example of custom callback:
https://github.com/microsoft/LightGBM/blob/aedfdd0d1e4ac570b539d17d74de1610acd099df/examples/python-guide/advanced_example.py#L196-L205
